src/kernel/thread/sync/mutex.c

```diff
@@ -54,17 +55,31 @@ int mutex_lock(struct mutex *m) {
 
 	errcheck = (m->attr.type == MUTEX_ERRORCHECK);
 
-	wait_ret = WAITQ_WAIT(&m->wq, ({
-		int done;
+#define trylock_code                                    \
+	({                                                  \
+	int done;                                           \
+	sched_lock();                                       \
+	ret = mutex_trylock(m);                             \
+	done = (ret == 0) || (errcheck && ret == -EDEADLK); \
+	if (!done)                                          \
+		mutex_priority_inherit(current, m);             \
+	sched_unlock();                                     \
+	done;                                               \
+	})
+
+	if (abstime == NULL) {
+		wait_ret = WAITQ_WAIT(&m->wq, trylock_code);
```

Comment:
You can use WAITQ_WAIT_TIMEOUT in both cases because
```
#define WAITQ_WAIT(wq, cond_expr) \
	WAITQ_WAIT_TIMEOUT(wq, cond_expr, SCHED_TIMEOUT_INFINITE)
```